google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITCommitTimestampTest.java

```diff
@@ -0,0 +1,319 @@
+    expectedException.expectMessage("Errors parsing Spanner DDL statement");
+    String statement = "ALTER TABLE T ALTER COLUMN T3 SET OPTIONS (allow_commit_timestamp=bogus)";
+    dbAdminClient
+        .updateDatabaseDdl(instanceId, databaseId, ImmutableList.of(statement), null)
+        .get();
+  }
+
+  @Test
+  public void invalidColumnType() throws Exception {
+    // error_catalog error OptionErrorList
+    expectedException.expect(isSpannerException(ErrorCode.FAILED_PRECONDITION));
+    expectedException.expectMessage("Option only allowed on TIMESTAMP columns");
+    String statement = "ALTER TABLE T ADD COLUMN T4 INT64 OPTIONS (allow_commit_timestamp=true)";
+    try {
+      dbAdminClient
+          .updateDatabaseDdl(instanceId, databaseId, ImmutableList.of(statement), null)
+          .get();
+    } catch (ExecutionException e) {
+      throw SpannerExceptionFactory.newSpannerException(e.getCause());
+    }
+  }
+
+  private void alterColumnOption(String databaseId, String table, String opt) throws Exception {
+    String statement =
+        "ALTER TABLE "
+            + table
+            + " ALTER COLUMN ts"
+            + " SET OPTIONS (allow_commit_timestamp="
+            + opt
+            + ")";
+    dbAdminClient
+        .updateDatabaseDdl(instanceId, databaseId, ImmutableList.of(statement), null)
+        .get();
+  }
+
+  private void writeAndVerify(DatabaseClient client, Timestamp ts) {
+    Timestamp commitTimestamp =
+        client.write(
+            Arrays.asList(
+                Mutation.newInsertOrUpdateBuilder("T1").set("ts").to(ts).build(),
+                Mutation.newInsertOrUpdateBuilder("T2").set("ts").to(ts).build(),
+                Mutation.newInsertOrUpdateBuilder("T3").set("ts").to(ts).build()));
+    if (ts == Value.COMMIT_TIMESTAMP) {
+      ts = commitTimestamp;
+    }
+    assertThat(readRow(client, "T1", Key.of(ts), "ts").getTimestamp(0)).isEqualTo(ts);
+    assertThat(readRow(client, "T2", Key.of(ts), "ts").getTimestamp(0)).isEqualTo(ts);
+    assertThat(readRow(client, "T3", Key.of(ts), "ts").getTimestamp(0)).isEqualTo(ts);
+  }
+
+  @Test
+  // 1) Write timestamps in the past
+  // 2) Set all interleaved tables allow_commmit_timestamp=true
+  // 3) Use commit timestamp in all tables
+  // 4) Set all interleaved tables allow_commmit_timestamp=null
+  // 5) Write timestamps in the future
+  public void interleavedTable() throws Exception {
+    Database db =
+        testHelper.createTestDatabase(
+            "CREATE TABLE T1 (ts TIMESTAMP) PRIMARY KEY (ts)",
+            "CREATE TABLE T2 (ts TIMESTAMP) PRIMARY KEY (ts), INTERLEAVE IN PARENT T1",
+            "CREATE TABLE T3 (ts TIMESTAMP) PRIMARY KEY (ts), INTERLEAVE IN PARENT T2");
+    DatabaseClient client = testHelper.getDatabaseClient(db);
+    String databaseId = db.getId().getDatabase();
+
+    Timestamp timeNow = Timestamp.ofTimeMicroseconds(Instant.now().toEpochMilli() * 1000);
```

Comment:
Using `TimeUnit.MICROSECONDS.convert(Instant.now().toEpochMilli(), TimeUnit.MILLISECONDS)` instead of `millis * 1000` reduces the chances of ever making a calculation mistake when converting time values between different units.

Also: Using the actual current system time makes this test case depend on the fact that the local system time will not be lagging behind the time of the backend. I think it would be better to use a fixed time here as well, or at least add a negative value to the current local system time to reduce the probability that a small discrepancy between the two would make the test case fail.

Reply:
I see your point, but this test has been successfully passing in current format since always. Would prefer keeping it this way.

Reply:
Ack.